src/main/jbake/content/blog/2016/12/19/hawkular-metrics-compression.adoc

```diff
@@ -0,0 +1,93 @@
+
+As we store values in a millisecond precision, it was important for us to increase the size of the block's maximum delta value. In Facebook's implementation they reserved 14 bits to represent the first delta from the block start (giving them 4 hour blocks with 1 second precision) where as we use 27 bits to get a maximum of 24 hour blocks with a millisecond precision. In 0.20.0/0.21.0 we however use 2 hour block sizes but with millisecond precision. We will revisit the block size requirements later.
+
+=== Insert pipeline
+
+We didn't change anything. When you issue a write to the Cassandra, we store it just like before to the `data` table. No restrictions on out-of-order times or changes to the persistence. Instead, we opted to do everything as an ETL (Extract Transform Load) job on the write path as described in the next section and merge results from both tables in the read path. 
+
+=== Compression job
+
+A while ago, we introduced a job scheduler to the Hawkular-Metrics and it's been improved on 0.20.0 and 0.21.0 based on our requirements to do reliable repeated compression jobs. When the server launches it schedules a repeated job called `CompressionData` that is ran every 2 hours at the start of odd hours. Should you for some reason miss one execution (because the server down for example), the scheduler will immediately start the missing jobs in order with only one being active at a time (cluster wide, state is maintained in Cassandra). The scheduled job `CompressionData` will take care of our ETL process for the whole cluster (this is something we will also improve later so that one can scale the amount of compression jobs, but for now we deemed it to be enough fast).
+
+When the job is started, the job itself takes a look at the intended runtime (irrelevant of the actual running time as that could be influenced by server downtime or previous job) and starts to fetch all the metrics that were created during the last 2 even hour block. For example, assuming the job should start at 03:00, we will compress the datapoints that were sent between 00:00 and 02:00. This allows up to one hour of out-of-order writes to be still compressed. If an out-of-order writes happens after this, it is kept uncompressed in the Cassandra and merged during the read phase.
```

Comment:
Go with "in Cassandra". There are several places where "in the Cassandra" should be replaced with "in Cassandra".